Resolves Rayleigh-Bénard Convection issue #1356 (#1403)
chaos-polymtl/lethe: doc/source/examples/multiphysics/rayleigh-benard-convection/rayleigh-benard-convection.rst

Both files mentioned below are located in the example's folder (``examples/multiphysics/rayleigh-benard-convection``).

- Parameter file for :math:`Ra=10^4`: ``rayleigh-benard-convection-Ra10k.prm``
- Parameter file for :math:`Ra=10^6`: ``rayleigh-benard-convection-Ra1M.prm``

-----------------------------
Description of the Case
-----------------------------

> In this example, we evaluate the performance of the ``lethe-fluid`` solver in the simulation of the stability of natural convection within a two-dimensional square domain. Our results are tested against the benchmark presented by Ouertatani et al. [#ouertatani]_ . The following schematic describes the geometry and dimensions of the simulation:

Reviewer: ```suggestion
In this example, we evaluate the performance of the ``lethe-fluid`` solver in the simulation of the stability of natural convection within a two-dimensional square domain. Our results are tested against the benchmark presented by Ouertatani *et al.* [#ouertatani]_  The following schematic describes the geometry and dimensions of the simulation:
```

.. note:: 
    All four boundary conditions for the velocity are ``noslip``. The sides wall are set to ``convection-radiation-flux`` with a heat transfer coefficient of 0 to have isolated walls. The top and bottom walls are set to a ``temperature`` to force a value of 0 and 10, respectively. 

--------------
Parameter File
--------------

> For simplicity, we present only the parameter file for :math:`Ra=10^4` only.

Reviewer: ```suggestion
For simplicity, we present only the parameter file for :math:`Ra=10^4`.
```

where :math:`\beta` and :math:`T_0` denote thermal expansion coefficient and a reference temperature, respectively.

A two-dimensional block of fluid is heated from its bottom wall at :math:`t = 0` s. The temperature of the bottom wall is equal to :math:`T_\text{h}=10`, the temperature of the top wall is equal to :math:`T_\text{c}=0`, and the left and right walls are insulated. By heating the fluid from the bottom wall, the buoyant force (natural convection) creates vortices inside the fluid. The shape and count of these vortices depend on the Rayleigh number and the Prandtl dimensionless numbers [#ouertatani]_ [#venturi2010]_ :

.. math::
  Ra &= \frac{\rho^2 \beta g (T_\text{h} - T_\text{c}) H^3 c_\text{p}}{k \mu} \\
  Pr &= \frac{\mu c_\text{p}}{k}

where :math:`\rho` is the fluid density, :math:`g` is the magnitude of gravitational acceleration, :math:`H` denotes the characteristic length, :math:`k` is the thermal conduction coefficient, :math:`\mu` is the dynamic viscosity, and :math:`c_\text{p}` is the specific heat capacity.

> In this example, we simulate the Rayleigh-Bénard convection problem at two Rayleigh numbers ( :math:`Ra=10^4` and :math:`Ra=10^6` ) with a Prandtl number of :math:`Pr=0.71` which correspond to air. According to the literature [#ouertatani]_ , we should see one big convective cell at steady-state for both :math:`Ra=10^4` and :math:`Ra=10^6`, but for the latter, there should also be two small voriticies in opposite corners rotating in the reverse direction of the big vortex. For simplicity, the gravity magnitude is set to 10 pointing in the opposite direction of the y-axis for both simulations. Additionally, because the two dimensionless numbers above are the only thing that characterize the flow we may choose the remaining parameters as we want. Here we chose to fix :math:`\rho = 1`, :math:`H = 1`, :math:`c_\text{p} = 100` and :math:`\mu = 0.071`. Thus, the Rayleigh number is controlled only by the thermal expansion coefficient (:math:`\beta = 0.71` or :math:`\beta = 71`). In other words, we change the Rayleigh number by changing the thermal expansion coefficient of the fluid.

Reviewer 2: ```suggestion
In this example, we simulate the Rayleigh-Bénard convection problem at two Rayleigh numbers ( :math:`Ra=10^4` and :math:`Ra=10^6` ) with a Prandtl number of :math:`Pr=0.71` which correspond to air. According to the literature [#ouertatani]_ , we should see one big convective cell at steady-state for both :math:`Ra=10^4` and :math:`Ra=10^6`, but for the latter, there should also be two small vortices in opposite corners rotating in the reverse direction of the big vortex. For simplicity, the gravity magnitude is set to 10 pointing in the opposite direction of the y-axis for both simulations. Additionally, because the two dimensionless numbers above are the only thing that characterize the flow we may choose the remaining parameters as we want. Here we chose to fix :math:`\rho = 1`, :math:`H = 1`, :math:`c_\text{p} = 100` and :math:`\mu = 0.071`. Thus, the Rayleigh number is controlled only by the thermal expansion coefficient (:math:`\beta = 0.71` or :math:`\beta = 71`).
```

.. note:: 

>     All four boundary conditions are ``noslip``.

Reviewer 2: What about the thermal boundary conditions? Maybe I would recall them here or otherwise I would say all four boundary conditions for the velocity are noslip

where :math:`\beta` and :math:`T_0` denote thermal expansion coefficient and a reference temperature, respectively.

> A two-dimensional block of fluid is heated from its bottom wall at :math:`t = 0` s. The temperature of the bottom wall is equal to :math:`T_\text{h}=10`, the temperature of the top wall is equal to :math:`T_\text{c}=0`, and the left and right walls are insulated. By heating the fluid from the bottom wall, the buoyant force (natural convection) creates vortices inside the fluid. The shape and number of these vortices depend on the Rayleigh number and the Prandtl [#ouertatani]_ [#venturi2010]_ :

Reviewer: ```suggestion
A two-dimensional block of fluid is heated from its bottom wall at :math:`t = 0` s. The temperature of the bottom wall is equal to :math:`T_\text{h}=10`, the temperature of the top wall is equal to :math:`T_\text{c}=0`, and the left and right walls are insulated. By heating the fluid from the bottom wall, the buoyant force (natural convection) creates vortices inside the fluid. The shape and count of these vortices depend on the Rayleigh and the Prandtl dimensionless numbers [#ouertatani]_ [#venturi2010]_ :
```

>     .. math::

Reviewer: Due to indentation, the equation are outlined as if they a citation or something (see capture)
![image](https://github.com/user-attachments/assets/cdc098a3-ed34-4637-8bdb-27ca354fb009)

To correct the indentations on lines 50-52:

> .. math::
>     Ra &= \frac{\rho^2 \beta g (T_\text{h} - T_\text{c}) H^3 c_\text{p}}{k \mu} \\
>     Pr &= \frac{\mu c_\text{p}}{k}

Reviewer: Can you also fix the indentation for lines 39-44 please, so that other people don't get the same issue if they reference to this example while making another:

> 
> The incompressible Navier-Stokes equations with a Boussinesq approximation for the buoyant force are:
> 
> .. math::
>     \nabla \cdot {\bf{u}} = 0
> 
> .. math::
>     \rho \frac{\partial {\bf{u}}}{\partial t} + \rho ({\bf{u}} \cdot \nabla) {\bf{u}} = -\nabla p + \nabla \cdot {\bf{\tau}} - \rho \beta {\bf{g}} (T - T_0)

      Ra &= \frac{\rho^2 \beta g (T_\text{h} - T_\text{c}) H^3 c_\text{p}}{k \mu} \\
      Pr &= \frac{\mu c_\text{p}}{k}

where :math:`\rho` is the fluid density, :math:`g` is the magnitude of gravitational acceleration, :math:`H` denotes the characteristic length, :math:`k` is the thermal conduction coefficient, :math:`\mu` is the dynamic viscosity, and :math:`c_\text{p}` is the specific heat capacity.

> In this example, we simulate the Rayleigh-Bénard convection problem at two Rayleigh numbers ( :math:`Ra=10^4` and :math:`Ra=10^6` ) with a Prandtl number of :math:`Pr=0.71` which correspond to air. According to the literature [#ouertatani]_ , we should see one big convective cell at steady-state for both :math:`Ra=10^4` and :math:`Ra=10^6`, but for the latter, there should also be two small voriticies in opposite corners rotating in the reverse direction of the big vortex. The gravity magnitude is set to -10 for both simulations for simplicity. Additionally, because the two adimensionale number above are the only thing that characterize the flow we may choose the remaining parameter as we want. Here we chose to fix :math:`\rho = 1`, :math:`H = 1`, :math:`c_\text{p} = 100` and :math:`\mu = 0.071`. Thus, the Rayleigh number is controlled only by the thermal expension coefficient (:math:`\beta = 0.71` or :math:`\beta = 71`). In other words, we change the Rayleigh number by changing the thermal expansion coefficient of the fluid.

Reviewer: ```suggestion
In this example, we simulate the Rayleigh-Bénard convection problem at two Rayleigh numbers ( :math:`Ra=10^4` and :math:`Ra=10^6` ) with a Prandtl number of :math:`Pr=0.71` which correspond to air. According to the literature [#ouertatani]_ , we should see one big convective cell at steady-state for both :math:`Ra=10^4` and :math:`Ra=10^6`, but for the latter, there should also be two small voriticies in opposite corners rotating in the reverse direction of the big vortex. For simplicity, the gravity magnitude is set to 10 pointing in the opposite direction of the y-axis for both simulations. Additionally, because the two dimensionless numbers above are the only thing that characterize the flow we may choose the remaining parameters as we want. Here we chose to fix :math:`\rho = 1`, :math:`H = 1`, :math:`c_\text{p} = 100` and :math:`\mu = 0.071`. Thus, the Rayleigh number is controlled only by the thermal expansion coefficient (:math:`\beta = 0.71` or :math:`\beta = 71`). In other words, we change the Rayleigh number by changing the thermal expansion coefficient of the fluid.
```

.. note:: 

>     All four boundary conditions are ``noslip``, and an external 

Reviewer: This note repeats what is said above on the gravity.

Parameter File
--------------

> For simplicity, we present only parameter file for :math:`Ra=10^4` only.

Reviewer: ```suggestion
For simplicity, we present only the parameter file for :math:`Ra=10^4`.
```

.. note::   

>     Note that the resolution (256x256) is set to match the one of the article results [#ouertatani]_ . If desired, you can choose to reduced to 7 or 6 ``initial refinement level`` of the prm file to reduce the simulation time without loosing too much accuracy.

Reviewer: ```suggestion
    Note that the resolution (256x256) is set to match the one of the article results [#ouertatani]_ . If desired, you can choose to reduced to 7 or 6 the ``initial refinement level`` in the parameter files to reduce the simulation time without loosing too much accuracy.
```

Simulation Control
~~~~~~~~~~~~~~~~~~

The time integration is handled by a 1st order backward differentiation scheme 
`(bdf1)`. A simulation time of :math:`24` s has been selected to ensure that the system reaches a quasi-steady state. The initial time-step is set to :math:`0.001` s, but this will be adapted during the simulation.

.. note::   
    This example uses an adaptive time-stepping method, where the 

>     time-step is modified during the simulation to keep the maximum value of the CFL condition below a given threshold (0.5 here). Using ``output control = time``, and ``output time frequency = 0.5`` the simulation results are written every 0.5 seconds regardless of the time steps.

Reviewer: ```suggestion
    time-step is modified during the simulation to keep the maximum value of the CFL condition below a given threshold (0.5 here). Using ``output control = time``, and ``output time frequency = 0.5`` the simulation results are written every 0.5 seconds regardless of the time-steps.
```

  mpirun -np 8 lethe-fluid rayleigh-benard-convection-Ra10k.prm

or

.. code-block:: text
  :class: copy-button

  mpirun -np 8 lethe-fluid rayleigh-benard-convection-Ra1M.prm

to run the simulations using eight CPU cores for the :math:`Ra=10^4` and :math:`Ra=10^6` cases respectively. Feel free to use more CPU if available. 

.. warning:: 
    Make sure to compile lethe in `Release` mode and 
    run in parallel using mpirun. The first simulation takes

>     :math:`\approx` 20 minutes on 8 processes and the second at :math:`Ra=10^6` can take more than 2 days.

Reviewer 2: Why does it take more than 2 days? is it because it takes a long time to converge? I would add a justification here.

Author: It converge in less simulation time (approximately 4 seconds), but the time step are much smaller to respect the CFL.

  mpirun -np 8 lethe-fluid rayleigh-benard-convection-Ra10k.prm

or

.. code-block:: text
  :class: copy-button

  mpirun -np 8 lethe-fluid rayleigh-benard-convection-Ra1M.prm

> to run the simulations using eight CPU cores at the two Rayleigh number. Feel free to use more CPU if available. 

Reviewer: ```suggestion
to run the simulations using eight CPU cores for the :math:`Ra=10^4` and :math:`Ra=10^6` cases respectively. Feel free to use more CPU if available. 
```

-------
Results
-------

The following animation shows the evolution of the temperature field with the flow direction for the simulation at :math:`Ra=10^6`:

.. raw:: html

    <iframe width="640" height="360" src="https://www.youtube.com/embed/NSJJpPauiXo" frameborder="0" allowfullscreen></iframe>

Below, we also present the velocity profiles at steady-state of our simulation compared to the ones presented by Ouertatani et al. [#ouertatani]_ as a verfification of the Lethe software. 

|fig1| |fig2|

.. |fig1| image:: images/solution-rayleigh-uy.png
    :width: 45%

.. |fig2| image:: images/solution-rayleigh-xv.png
    :width: 47%

The results can be postprocessed using the provided Python script (``rayleigh-benard-convection.py``). Here is an example of how to call the script:

.. code-block:: text
  :class: copy-button

  python3 rayleigh-benard-convection.py -f ./output_10k -f ./output_1M -Ra 10k -Ra 1M

This script extracts the velocity in the :math:`x` and :math:`y` directions at the mid-width (:math:`x=0.5`) and mid-height (:math:`y=0.5`) respectively and create the above plots.

.. warning::

>   The orientation of the vortex rotation obtained with the simulation may differ from the one above due to machine precision that generate the initial instability.

Reviewer: ```suggestion
  The orientation of the vortex rotation obtained with the simulation may differ from the one above due to machine precision that generates the initial instability.
```

  mpirun -np 8 lethe-fluid rayleigh-benard-convection-Ra10k.prm

or

.. code-block:: text
  :class: copy-button

  mpirun -np 8 lethe-fluid rayleigh-benard-convection-Ra1M.prm

to run the simulations using eight CPU cores at the two Rayleigh number. Feel free to use more CPU if available. 

.. warning:: 
    Make sure to compile lethe in `Release` mode and 
    run in parallel using mpirun. The first simulation takes
    :math:`\approx` 20 minutes on 8 processes and the second at :math:`Ra=10^6` can take more than 2 days.

-------
Results
-------

> The following animation shows the evolution of the temperature field with the flow direction for the simulation at :math:`Ra=10^6` :

Reviewer: ```suggestion
The following animation shows the evolution of the temperature field with the flow direction for the simulation at :math:`Ra=10^6`:
```
